### PR TITLE
GODRIVER-1793 Add tests for nil options to InsertOne

### PR DESCRIPTION
**Pull request by kevinAlbs**

### CODE REVIEW

File: mongo/collection.go [hunk 1]
```diff
@@ -326,15 +326,13 @@ func (coll *Collection) insert(ctx context.Context, documents []interface{},
...
 
-	imOpts := make([]*options.InsertManyOptions, len(opts))
-	for i, opt := range opts {
-		imo := options.InsertMany()
-		if opt != nil && opt.BypassDocumentValidation != nil && *opt.BypassDocumentValidation {
-			imo = imo.SetBypassDocumentValidation(*opt.BypassDocumentValidation)
-		}
-		imOpts[i] = imo
+	ioOpts := options.MergeInsertOneOptions(opts...)
+	var imOpts options.InsertManyOptions
```
Comment (divjotarora) [inline mongo/collection.go:330]: This can be declared as `imOpts := &options.InsertMany()`. For the case of `InsertMany`, this is probably a no-op, but in other cases, the options constructors can be responsible for setting defaults, so this is the idiomatic way to declare options.

File: mongo/integration/collection_test.go [hunk 2]
```diff
@@ -74,6 +74,37 @@ func TestCollection(t *testing.T) {
 			assert.True(mt, ok, "expected error type %v, got %v", mongo.WriteException{}, err)
 			assert.NotNil(mt, we.WriteConcernError, "expected write concern error, got %+v", we)
 		})
+
+		// Require 3.2 servers for bypassDocumentValidation support.
+		nilOptsOpts := mtest.NewOptions().MinServerVersion("3.2")
+		mt.RunOpts("nil opts", nilOptsOpts, func(mt *mtest.T) {
+			assertOptSet := func(expectSet bool) {
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:81]: This might be better organized/readable as a set of table-driven tests? If you have this sort of local helper function that gets called for different configurations in the test, it's usually a sign that things can be refactored into a table that runs sub-tests.

File: mongo/collection.go [hunk 3]
```diff
@@ -326,15 +326,13 @@ func (coll *Collection) insert(ctx context.Context, documents []interface{},
...
 	opts ...*options.InsertOneOptions) (*InsertOneResult, error) {
 
-	imOpts := make([]*options.InsertManyOptions, len(opts))
-	for i, opt := range opts {
-		imo := options.InsertMany()
-		if opt != nil && opt.BypassDocumentValidation != nil && *opt.BypassDocumentValidation {
-			imo = imo.SetBypassDocumentValidation(*opt.BypassDocumentValidation)
-		}
-		imOpts[i] = imo
+	ioOpts := options.MergeInsertOneOptions(opts...)
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply mongo/collection.go:329]: 🤦 forgot to publish this on initial review...

File: mongo/collection.go (hunk 1 above)
Comment (kevinAlbs) [reply mongo/collection.go]: Good point, done.

File: mongo/integration/collection_test.go (hunk 2 above)
Comment (kevinAlbs) [reply mongo/integration/collection_test.go]: Oh good idea... I just learned that this was a thing outside of the Go driver from the testing talk in the onboarding doc.

File: mongo/integration/collection_test.go [hunk 4]
```diff
@@ -74,6 +74,48 @@ func TestCollection(t *testing.T) {
 			assert.True(mt, ok, "expected error type %v, got %v", mongo.WriteException{}, err)
 			assert.NotNil(mt, we.WriteConcernError, "expected write concern error, got %+v", we)
 		})
+
+		// Require 3.2 servers for bypassDocumentValidation support.
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:78]: There should be another `mt.Run`/`mt.RunOpts` call here so all of the table tests are run inside a larger context. Right now, the tests would be named something like `TestCollection/insert_one/only_nil_is_passed`. If we add another level of nesting, it'd be `TestCollection/insert_one/options_are_converted/only_nil_is_passed`.

Comment (kevinAlbs) [reply mongo/integration/collection_test.go:78]: Done.

File: mongo/integration/collection_test.go [hunk 5]
```diff
@@ -74,6 +74,50 @@ func TestCollection(t *testing.T) {
 			assert.True(mt, ok, "expected error type %v, got %v", mongo.WriteException{}, err)
 			assert.NotNil(mt, we.WriteConcernError, "expected write concern error, got %+v", we)
 		})
+
+		mt.Run("options are converted", func(mt *mtest.T) {
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:78]: nit: Skip this top level test for server versions < 3.2 rather than skipping all sub-tests individually.

File: mongo/integration/collection_test.go [hunk 6]
```diff
@@ -74,6 +74,50 @@ func TestCollection(t *testing.T) {
...
+					_, err := mt.Coll.InsertOne(mtest.Background, doc, testCase.opts...)
+					assert.Nil(mt, err, "InsertOne error: %v", err)
+					optName := "bypassDocumentValidation"
+					evt := mt.GetStartedEvent()
+
+					val, err := evt.Command.LookupErr(optName)
+					if testCase.expectOptionSet {
+						assert.Nil(mt, err, "expected %v to be set but got: %v", optName, err)
+						assert.True(mt, val.Boolean(), "expected %v to be true but got: %v", optName, val.Boolean())
+					} else {
```
Comment (divjotarora) [inline mongo/integration/collection_test.go:115]: nit: return early from the previous `if` statement instead of using `else`. There's no important reason for doing this AFAIK. It just aligns well with other tests we've written.

**Review by divjotarora — APPROVED** [review]
Minor nits in the test file, but LGTM.

**Review by iwysiu — APPROVED** [review]
lgtm with divjot's nits

File: mongo/integration/collection_test.go (hunk 6 above)
Comment (kevinAlbs) [reply mongo/integration/collection_test.go]: Done.

File: mongo/integration/collection_test.go (hunk 5 above)
Comment (kevinAlbs) [reply mongo/integration/collection_test.go]: Done.

**Review by kevinAlbs — COMMENTED** [review]
Will merge after evergreen runs on latest commit.
